File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l= 0
        max_prof = 0
        for r in range(1,len(prices)):
            if prices[l]<prices[r]:
                profit = prices[r]-prices[l]
                max_prof = max(max_prof,profit)

            else:
                l=r
            

        return max_prof
        

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.java
class Solution {
    Map<String, Integer> dp = new HashMap<>();
    public int maxProfit(int[] prices) {
        return dfs( prices, true, 0);
    }
    public int dfs(int[] prices, boolean buy, int i){
         String key = i + "-" + buy;
        if (dp.containsKey(key)) {
            return dp.get(key);
        }
        
        if(i>=prices.length){
            return 0;
        }
        
        //skipping buying or selling on that day
        int cooldown = dfs(prices, buy, i+1 );

        
        if(buy==true){
            //if i buy coin that day
            int bcost = dfs(prices, false, i+1)-prices[i];
            
            dp.put(key, Math.max(bcost, cooldown));

        }

        
        else{
            //if i sell coin that day
            int scost = dfs(prices, true, i+2)+prices[i];
            
            dp.put(key, Math.max(scost, cooldown));
        }
        
        return dp.get(key);

        
    }
}
